### DISCUSSION TITLE
mpremote for esp32 with MICROPY_PREVIEW_VERSION_2

### DISCUSSION THREAD

**Discussion by shariltumin** [micropython/micropython#13663] [Libraries & Drivers]
The mpremote allows us to execute files from the unix directory from within micropython. This is good because we do not have to copy files to flash during development, which extends the life of the flash.

However, compiling ESP32 with:

```
make V=1 BOARD=ESP32 MICROPY_PREVIEW_VERSION_2=1
```

will not include mount and umount methods under the os module. The MP2 version uses vfs modules for mount and umount.

We need to compile ESP32 with:

```
make V=1 BOARD=ESP32 MICROPY_PREVIEW_VERSION_2=1 MICROPY_PY_OS=1 MICROPY_VFS_LFS2=1

```
Now with mpremote-1.22.0

```sh
$ mpremote u0 mount .
Traceback (most recent call last):
  File "/home/sharil/opt/bin/mpremote", line 8, in <module>
    sys.exit(main())
  File "/home/sharil/opt/lib/python3.10/site-packages/mpremote/main.py", line 528, in main
    handler_func(state, args)
  File "/home/sharil/opt/lib/python3.10/site-packages/mpremote/commands.py", line 226, in do_mount
    state.transport.mount_local(path, unsafe_links=args.unsafe_links)
  File "/home/sharil/opt/lib/python3.10/site-packages/mpremote/transport_serial.py", line 508, in mount_local
    self.exec("__mount()")
  File "/home/sharil/opt/lib/python3.10/site-packages/mpremote/transport_serial.py", line 287, in exec
    raise TransportError("exception", ret, ret_err)
mpremote.transport.TransportError: ('exception', b'', b'Traceback (most recent call last):\r\n  File "<stdin>", line 1, in <module>\r\n  File "<stdin>", line 272, in __mount\r\nAttributeError: \'module\' object has no attribute \'mount\'\r\n')
```
So we need to modify mpremote for ESP32 compile for MICROPY_PREVIEW_VERSION_2 with MICROPY_PY_OS.

```sh
$ diff transport_serial.py transport_serial.py-ORIG
600c600
<             self.exec('vfs.umount("/remote")')
---
>             self.exec('os.umount("/remote")')
621c621
< import vfs, os, io, struct, micropython
---
> import os, io, struct, micropython
946c946
<     vfs.mount(RemoteFS(RemoteCommand()), '/remote')
---
>     os.mount(RemoteFS(RemoteCommand()), '/remote')
```

The easiest way is to copy mpremote to espremote and edit the transport_serial.py file in the new espremote directory. Also copy the mpremote start file to espremote and edit it to use espremote.main file. ```from espremote.main import main```

```sh
$ espremote u0 mount .
Local directory . is mounted at /remote
Connected to MicroPython at /dev/ttyUSB0
Use Ctrl-] or Ctrl-x to exit this shell
>
```
This is just for information for those compiling their own firmware. Hopefully it will be useful to someone.

**Comment by jimmo** [comment]
Thanks @shariltumin -- Indeed this was (briefly) discussed in the PR that moved the methods. https://github.com/micropython/micropython/pull/13584#issuecomment-1926101241

We'll need to address this soon, and it'll be a good opportunity to make `mpremote` more flexible.

**Reply by Josverl** [reply]
I'd prefer to have one mpremote to connect to all ports and MP1. x as well as MP2.X

diversing into different tools for different ports would be a setback, IMHO.

As there is a efficient way to detect a targets board capability, I think that is the better option

**Reply by shariltumin** [reply]
I only need mpremote to work with esp32 **now**, so that is why.

My suggestion is only for the interim period. I completely agree with you. I am confident that the CORE team will provide us with a single mpremote that covers all MCUs.
